### PR TITLE
Implementation of the nf-core YaHS module

### PR DESCRIPTION
**Pull request by ksenia-krasheninnikova**
## PR checklist

Closes [#2569](https://github.com/nf-core/modules/issues/2569) 

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [ ] Emit the `versions.yml` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

File: modules/nf-core/yahs/main.nf [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+    conda (params.enable_conda ? "bioconda::yahs=1.2a.2" : null)
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/yahs:1.2a.2--h7132678_0':
+        'quay.io/biocontainers/yahs:1.2a.2--h7132678_0' }"
+
+    input:
+    tuple val(meta), path(bed)
+    tuple path(ref), path(fai)
+    val(if_break)
```
Comment (pinin4fjords) [inline modules/nf-core/yahs/main.nf:13]: I think, according to [current guidelines](https://nf-co.re/docs/contributing/modules#general), that these should all go through `ext.args`.

File: modules/nf-core/yahs/main.nf [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    tuple val(meta), path("yahs.out_scaffolds_final.agp"), emit: scaffolds_agp
+    tuple val(meta), path("yahs.out.bin"), emit: binary
+    path "versions.yml"           , emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    def break_param = if_break ? '' : ' --no-contig-ec'
```
Comment (pinin4fjords) [inline modules/nf-core/yahs/main.nf:28]: As referenced above, I think all the options here can come directly through from ext.args without the need for explicit `val` inputs.

File: modules/nf-core/yahs/main.nf [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    def motif_param = motif ? ' -e ${motif} ' : ''
+    def resolutions_param = resolutions ? ' -r ${resolutions} ' : ''
+    """
+    yahs ${break_param} \\
+        ${motif_param} \\
+        ${resolutions_param} \\
+        $args \\
+        $ref \\
+        $bed > out.log
```
Comment (pinin4fjords) [inline modules/nf-core/yahs/main.nf:37]: You shouldn't redirect STDOUT, Nextflow will use that to populate .command.out.

File: modules/nf-core/yahs/meta.yml [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+      type: file
+      description: FASTA reference file
+      pattern: "*.{fasta,fa}"
+
+  - fai:
+      type: file
+      description: index of the reference file
+      pattern: "*.{fai}"
+
+  - if_break:
```
Comment (pinin4fjords) [inline modules/nf-core/yahs/meta.yml:38]: Unless my points on `val` inputs are contradicted by someone else, the related documentation will also need to be removed (we defer to the upstream documentation).

File: modules/nf-core/yahs/main.nf [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+    input:
+    tuple val(meta), path(bed)
+    tuple path(ref), path(fai)
+    val(if_break)
+    val(motif)
+    val(resolutions)
+
+    output:
+    tuple val(meta), path("yahs.out_scaffolds_final.fa"), emit: scaffolds_fasta
```
Comment (pinin4fjords) [inline modules/nf-core/yahs/main.nf:18]: Could you vertically align the `emit:`s for readability? Have a look at other modules to see what I mean.

File: modules/nf-core/yahs/meta.yml [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test' ]
+  #
+  - bed:
+      type: file
+      description: BED file containing coordinates of duplicate alignments
+      pattern: "*.{bed}"
+
+  - fasta:
```
Comment (pinin4fjords) [inline modules/nf-core/yahs/meta.yml:28]: This should be `ref`, no?

File: tests/config/test_data.config [hunk 7]
```diff
@@ -549,5 +549,10 @@ params {
                 pretext = "${test_data_dir}/genomics/eukaryotes/galaxea_fascicularis/hic/jaGalFasc40_2.pretext"
             }
         }
+        'yahs' {
+            ilEupCent1_sorted_bed = "https://darwin.cog.sanger.ac.uk/sanger-tol-genomeassembly/test/yahs/ilEupCent1.sorted.bed"
```
Comment (pinin4fjords) [inline tests/config/test_data.config:553]: You should use existing data from the [nf-core test data repository](https://nf-co.re/docs/contributing/test_data_guidelines), or else PR some new (small) test files if there's nothing suitable.

These will be used in CI etc, and we don't want to be hammering the Sanger servers repeatedly.

**Review by pinin4fjords — CHANGES_REQUESTED** [review]

**Review by nvnieuwk — CHANGES_REQUESTED** [review]
Good job! Here are some suggestions :)

File: modules/nf-core/yahs/main.nf [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    input:
+    tuple val(meta), path(bed)
+    path fasta
+    path fai
+
+    output:
+    tuple val(meta), path("*scaffolds_final.fa") , emit: scaffolds_fasta
+    tuple val(meta), path("*scaffolds_final.agp"), emit: scaffolds_agp
+    tuple val(meta), path("*bin")                , emit: binary
+    path "versions.yml"                                  , emit: versions
```
Comment (pinin4fjords) [inline modules/nf-core/yahs/main.nf:19]: re-align ;-)

File: modules/nf-core/yahs/main.nf [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
```
Comment (pinin4fjords) [inline modules/nf-core/yahs/main.nf:25]: @nvnieuwk meant that you needed to pass task.ext.prefix like [this](https://github.com/nf-core/modules/blob/ee80d14721e76e2e079103b8dcd5d57129e584ba/modules/nf-core/gsea/gsea/main.nf#L45) and [this](https://github.com/nf-core/modules/blob/ee80d14721e76e2e079103b8dcd5d57129e584ba/tests/modules/nf-core/gsea/gsea/nextflow.config#L15).

**Review by pinin4fjords — CHANGES_REQUESTED** [review]
Almost there!

File: modules/nf-core/yahs/main.nf (hunk 9 above)
Comment (nvnieuwk) [reply modules/nf-core/yahs/main.nf:25]: Yes, my bad for not being clear enough about it :D

Comment (pinin4fjords) [reply modules/nf-core/yahs/main.nf:25]: @ksenia-krasheninnikova that might mean passing the prefix to the tool to determine output file names, or else renaming the files afterwards (as I had to do in the GSEA case).

Comment (ksenia-krasheninnikova) [reply modules/nf-core/yahs/main.nf:25]: Thank you both! I hope the last update does it.

**Review by nvnieuwk — APPROVED** [review]
LGTM! if tests pass

File: modules/nf-core/yahs/meta.yml [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+  - scaffolds_fasta:
+      type: file
+      description: FASTA file with resulting contigs
+      pattern: "yahs.out_scaffolds_final.fa"
+
+  - scaffolds_agp:
+      type: file
+      description: AGP file containing contigs placing coordinates
+      pattern: "yahs.out_scaffolds_final.agp"
```
Comment (nvnieuwk) [inline modules/nf-core/yahs/meta.yml:59]: Can you also update the pattern here?

File: modules/nf-core/yahs/main.nf [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+    input:
+    tuple val(meta), path(bed)
+    path fasta
+    path fai
+
+    output:
+    tuple val(meta), path("*scaffolds_final.fa") , emit: scaffolds_fasta
+    tuple val(meta), path("*scaffolds_final.agp"), emit: scaffolds_agp
+    tuple val(meta), path("*bin")                , emit: binary
```
Comment (nvnieuwk) [inline modules/nf-core/yahs/main.nf:18]: The output files aren't being detected, are you sure this can find them all?

Comment (pinin4fjords) [reply modules/nf-core/yahs/main.nf:18]: I think the test.yml just needs updating for the new prefix ('test')

File: modules/nf-core/yahs/meta.yml [hunk 12]
```diff
@@ -51,19 +51,19 @@ output:
   - scaffolds_fasta:
       type: file
       description: FASTA file with resulting contigs
-      pattern: "*scaffolds_final.fa"
+      pattern: "test_scaffolds_final.fa"
```
Comment (pinin4fjords) [inline modules/nf-core/yahs/meta.yml:54]: No, you were right before- these should hold the wildcards. It's test.yml needs fixing.

**Review by pinin4fjords — APPROVED** [review]
Looking good - thanks for addressing the feedback.

**Comment by ksenia-krasheninnikova** [comment]
Thank you for help @pinin4fjords and @nvnieuwk!

**Pull request merged**